KAFKA-19480: Recreate /migration when it has null value

harmadasg opened the pull request:
When using the [zookeeper-security-migration](https://kafka.apache.org/39/documentation.html#zk_authz_migration) tool without the '–enable.path.check' option, the script not only updates the ACLs for the existing znodes, but also creates any non-existing ones (with the ACL options specified) using null values based on the list defined in [ZkData.SecureRootPaths.](https://github.com/apache/kafka/blob/3.9/core/src/main/scala/kafka/zk/ZkData.scala#L1089-L1102) This is especially problematic for the /migration znode as the current logic only checks for the existence of the znode and later the migration process will hang when it tries to parse the null value over and over again. 

In summary, the migration cannot be completed if the zookeeper-security-migration script was run previously, and the only workaround is to manually remove the /migration znode in such cases. I propose a simple fix to circumvent the manual step by recreating the /migration znode if it contains a null value.

harmadasg committed:
KAFKA-19480: Recreate /migration when it has null value

viktorsomogyi commented:
@harmadasg when the node is originally created, what if we created them with an empty array instead of `null`? Or at least this one? It may simplify the PR and the logic.

harmadasg commented:
@viktorsomogyi thanks for the review,

 About creating the missing znodes, the logic iterates over [ZkData.SecureRootPaths](https://github.com/apache/kafka/blob/3.9/core/src/main/scala/kafka/zk/ZkData.scala#L1089-L1102) and calls `setACL` on every znode in the list ([here](https://github.com/apache/kafka/blob/b7b2676f5345b1683db1599a2c0d04461619ecbf/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L185) and [here](https://github.com/apache/kafka/blob/b7b2676f5345b1683db1599a2c0d04461619ecbf/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L195)).  Before that it calls [zkClient.makeSurePersistentPathExists(path)](https://github.com/apache/kafka/blob/b7b2676f5345b1683db1599a2c0d04461619ecbf/core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala#L271) to create the znode if it doesn't exists (with null value) which causes the issue.

Regarding creating the node with an empty array or skipping the znode altogether, I think this solution can work (like in [KAFKA-9267](https://issues.apache.org/jira/browse/KAFKA-9267) for the controller znode) but this wouldn't fix the null znode issue for people who ran [zookeeper-security-migration](https://kafka.apache.org/39/documentation.html#zk_authz_migration) tool in the past (from 3.4.1 to 3.9.1) and have already a null `migration` znode

Edit: I have added a new commit [352c6af](https://github.com/apache/kafka/pull/20128/commits/352c6afb3b343fdb8cc48489b7d112dec79943c5) to skip creating the migration znode in the zookeeper-security-migration tool

harmadasg committed:
KAFKA-19480: Do not create migration znode with null value in zookeeper-security-migration tool

viktorsomogyi commented:
@harmadasg thanks for the reference to KAFKA-9267, it strengthens the point that a default `null` could cause issues elsewhere. I think that users who'd upgrade to fix this NPE during a migration is slim and the workaround is easy, so I'd be in favor of changing the default to an empty array.

mumrah commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1734,20 +1734,27 @@ class KafkaZkClient private[zk] (
+          case Some(data) =>
+            MigrationZNode.decode(data, getDataResponse.stat.getVersion, getDataResponse.stat.getMtime)
+          case None =>
+            createInitialMigrationState(initialState, removeFirst = true)
```
Can we log a message here that we are fixing the `/migration` znode?

harmadasg replied:
fixed it in [809d34c](https://github.com/apache/kafka/pull/20128/commits/809d34c24244068fddc6ead90a54d785cf5b9206)

mumrah commented on core/src/main/scala/kafka/zk/KafkaZkClient.scala:
```diff
@@ -1734,20 +1734,27 @@ class KafkaZkClient private[zk] (
       CreateMode.PERSISTENT)
-    val response = retryRequestUntilConnected(createRequest)
+    val deleteOp = DeleteOp(MigrationZNode.path, ZkVersion.MatchAnyVersion)
+    val multi = MultiRequest((if (removeFirst) Some(deleteOp) else None).toSeq ++ Seq(createOp))
```
Similar to the co-pilot suggestion, we can simplify this to something like

```
var multi = if (removeFirst) {
  MultiRequest(deleteOp, createOp)
} else {
  MultiRequest(createOp)
}

harmadasg replied:
updated the code based on the suggestion

mumrah reviewed (commented):
Thanks for the patch @harmadasg! Overall this looks good to me, just a couple of comments. Once this is merged we should also update the 3.9 docs to mention the problem with the zk security migration tool as well as a workaround for 3.9.1 and older.

harmadasg committed:
KAFKA-19480: Fix review items

harmadasg committed:
KAFKA-19480: add workaround in docs

harmadasg commented:
> @harmadasg thanks for the reference to [KAFKA-9267](https://issues.apache.org/jira/browse/KAFKA-9267), it strengthens the point that a default `null` could cause issues elsewhere. I think that users who'd upgrade to fix this NPE during a migration is slim and the workaround is easy, so I'd be in favor of changing the default to an empty array.

@viktorsomogyi as discussed I only skipped processing /migration in `ZkSecurityMigrator ` as `zkClient.makeSurePersistentPathExists` is used in many places and I wanted to reduce the scope of my change. Also probably the empty array would still result in errors when trying to decode against a specific znode structure.

harmadasg commented:
@mumrah thanks for the review, I have addressed your points in the latest commits. I also opened https://github.com/apache/kafka-site/pull/708 to update the docs for the older releases

viktorsomogyi commented:
From my side it looks good (I don't insist on addressing the nits from copilot). Restarted the build, so let's wait for a green. This gives time to @mumrah if he has any more comments.

harmadasg committed:
KAFKA-19480: clarify workaround steps mentioning the name of the tool

harmadasg commented:
Rebased on latest 3.9 to retrigger CI

viktorsomogyi commented:
Closed in favor of #20627.
